storage/internal/witness.go

```diff
@@ -0,0 +1,242 @@
+// Copyright 2025 The Tessera authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"slices"
+	"strconv"
+	"strings"
+	"sync"
+
+	tessera "github.com/transparency-dev/trillian-tessera"
+	"github.com/transparency-dev/trillian-tessera/internal/parse"
+	"k8s.io/klog/v2"
+)
+
+// ProofFetchFn is the signature of a function that can return the consistency proof
+// for append-only between the two given tree sizes.
+type ProofFetchFn func(ctx context.Context, from, to uint64) [][]byte
+
+// NewWitnessGateway returns a WitnessGateway that will send out new checkpoints to witnesses
+// in the group, and will ensure that the policy is satisfied before returning. All outbound
+// requests will be done using the given client.
+func NewWitnessGateway(group tessera.WitnessGroup, client *http.Client, fetchProof ProofFetchFn) WitnessGateway {
+	urls := group.URLs()
+	slices.Sort(urls)
+	urls = slices.Compact(urls)
+	witnesses := make([]witness, len(urls))
+	postFnImpl := func(ctx context.Context, url string, body string) (pr postResponse, err error) {
+		req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, strings.NewReader(body))
+		if err != nil {
+			return postResponse{}, err
+		}
+		httpResp, err := client.Do(req)
+		if err != nil {
+			return postResponse{}, err
+		}
+		rb, err := io.ReadAll(httpResp.Body)
+		if err != nil {
+			return postResponse{}, fmt.Errorf("failed to read response body: %v", err)
+		}
+		_ = httpResp.Body.Close()
+		return postResponse{
+			statusCode: httpResp.StatusCode,
+			body:       rb,
+			headers:    httpResp.Header,
+		}, nil
+	}
+	for _, u := range urls {
+		witnesses = append(witnesses, witness{
+			url:        u,
+			size:       0,
+			post:       postFnImpl,
```

Comment:
It kinda feels like this func should have the url baked in here, and the func signature be something like `func(context.Context, string) (...)`, or will the `witness` ever need to change `url` before passing it into `post`?

Reply:
If this was an external API then I'd agree more strongly. I just tried changing it and the soft blocker is that if we lost the URL then the witness has no identifier, which makes any errors it returns ambiguous. Keeping the URL there allows us to say which witness failed.